api/filters/reference_instance_filter.go

```diff
@@ -0,0 +1,251 @@
+	callerTenantID := gjson.GetBytes(req.Body, path).String()
+	log.C(ctx).Infof("Reference Instance Update passed successfully. Called by \"%s\", instanceID: \"%s\"", callerTenantID, resourceID)
+	return next.Handle(req)
+}
+
+func (rif *referenceInstanceFilter) isReferencedShared(ctx context.Context, referencedInstanceID string) (bool, error) {
+	byID := query.ByField(query.EqualsOperator, "id", referencedInstanceID)
+	dbReferencedObject, err := rif.repository.Get(ctx, types.ServiceInstanceType, byID)
+	if err != nil {
+		return false, util.HandleStorageError(err, types.ServiceInstanceType.String())
+	}
+	referencedInstance := dbReferencedObject.(*types.ServiceInstance)
+
+	if !*referencedInstance.Shared {
+		return false, util.HandleInstanceSharingError(util.ErrReferencedInstanceNotShared, referencedInstanceID)
+	}
+	return true, nil
+}
+
+func (rif *referenceInstanceFilter) isValidPatchRequest(req *web.Request, instance *types.ServiceInstance) (bool, error) {
+	newPlanID := gjson.GetBytes(req.Body, planIDProperty).String()
+	if instance.ServicePlanID != newPlanID {
+		return false, util.HandleInstanceSharingError(util.ErrChangingPlanOfReferenceInstance, instance.ID)
+	}
+
+	parametersRaw := gjson.GetBytes(req.Body, "parameters").Raw
+	if parametersRaw != "" {
+		return false, util.HandleInstanceSharingError(util.ErrChangingParametersOfReferenceInstance, instance.ID)
+	}
+
+	return true, nil
+}
+
+func (rif *referenceInstanceFilter) isReferencePlan(ctx context.Context, servicePlanID string) (bool, error) {
+	dbPlanObject, err := rif.getObjectByID(ctx, types.ServicePlanType, servicePlanID)
+	if err != nil {
+		return false, err
+	}
+	plan := dbPlanObject.(*types.ServicePlan)
+	return plan.Name == instance_sharing.ReferencePlanName, nil
+}
+
+func (rif *referenceInstanceFilter) getObjectByID(ctx context.Context, objectType types.ObjectType, resourceID string) (types.Object, error) {
+	byID := query.ByField(query.EqualsOperator, "id", resourceID)
+	dbObject, err := rif.repository.Get(ctx, objectType, byID)
+	if err != nil {
+		return nil, util.HandleStorageError(err, objectType.String())
+	}
+	return dbObject, nil
+}
+
+func (rif *referenceInstanceFilter) validateOwnership(req *web.Request) error {
+	ctx := req.Context()
+	contextPath := fmt.Sprintf("%s.%s", contextKey, rif.tenantIdentifier)
+	callerTenantID := gjson.GetBytes(req.Body, contextPath).String()
+	referencedInstancePath := fmt.Sprintf("parameters.%s", instance_sharing.ReferencedInstanceIDKey)
+	referencedInstanceID := gjson.GetBytes(req.Body, referencedInstancePath).String()
+	byID := query.ByField(query.EqualsOperator, "id", referencedInstanceID)
+	dbReferencedObject, err := rif.repository.Get(ctx, types.ServiceInstanceType, byID)
+	if err != nil {
+		return util.HandleStorageError(err, types.ServiceInstanceType.String())
+	}
+	instance := dbReferencedObject.(*types.ServiceInstance)
+	sharedInstanceTenantID := instance.Labels[rif.tenantIdentifier][0]
```

Comment:
wouldn't this panic in case multitenancy is not enabled?

Reply:
Hey Eyal,

We wrap the function with a condition of existing "callerTenantID", 
which as far as I know, is being added to the request on the multi-tenant flow

```
	if callerTenantID != "" {
		err = rif.validateOwnership(req)
		if err != nil {
			return nil, err
		}
	}
```

Please let me know if I am wrong here with the flow of the multi-tenant.

Anyway, regarding the topic in general, we will try to go over things again and catch scenarios that can cause panics.

Thanks!

Reply:
1. if you extract the tenant ID from the body before calling this function why do you do it again and not just pass it in?
2. wouldn't the right way to check if multi-tenancy is enabled be `if rif.tenantIdentifier != ""`?

Reply:
Thanks, code updated